subsystems/mvdm/ntvdm/bios/vidbios.c

```diff
@@ -2942,6 +2942,295 @@ static VOID VidBiosPrintCharacter(CHAR Character, BYTE Attribute, BOOLEAN UseAtt
+                sizeof(BYTE));
+            break;
+        }
+
+        default:
+        {
+            DPRINT1("VidBiosSetPixel - unsupported video mode %X\n", Bda->VideoMode);
+            break;
+        }
+    }
+}
+
+static BYTE VidBiosGetPixel(WORD X, WORD Y, BYTE Page)
+{
+    SHORT ScreenWidth = VgaGetDisplayResolution().X;
+
+    if (IS_TEXT_MODE(Bda->VideoMode))
+    {
+        DPRINT1("Note, trying to access pixel (%d,%d) @ page %d in text mode %X - returning 0.\n", X, Y, Page, Bda->VideoMode);
+        return 0;
+    }
+
+    switch (Bda->VideoMode)
+    {
+        case 0x04: /* CGA 320x200 4c */
+        case 0x05: /* CGA 320x200 4c */
+        case 0x06: /* CGA 640x200 2c */
+        {
+            ULONG Address;
+            BYTE MemoryValue;
+            BYTE MaxColor;
+            BYTE ValueShift;
+            BYTE PixelsPerByte;
+
+            if (Bda->VideoMode == 0x06)
+            {
+                MaxColor = 1;
+                ValueShift = 7 - X % 8;
+                PixelsPerByte = 8;
+            }
+            else
+            {
+                MaxColor = 3;
+                ValueShift = (3 - X % 4) * 2;
+                PixelsPerByte = 4;
+            }
+
+            Address = TO_LINEAR((Y & 1) ? CGA_ODD_VIDEO_SEG : CGA_EVEN_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + ((Y / 2) * ScreenWidth + X) / PixelsPerByte);
+
+            EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+            return (MemoryValue >> ValueShift) & MaxColor;
+        }
+
+        case 0x0D: /* EGA 320x200 16c */
+        case 0x0E: /* EGA 640x200 16c */
+        case 0x10: /* EGA 640x350 16c */
+        case 0x12: /* VGA 640x480 16c */
+        {
+            ULONG Address;
+            BYTE OldRead;
+            BYTE Shift = 7 - X % 8;
+            BYTE Value = 0;
+            BYTE CurrentPlane;
+
+            Address = TO_LINEAR(GRAPHICS_VIDEO_SEG,
+                                Page * Bda->VideoPageSize + (Y * ScreenWidth + X) / 8);
+
+            /* Save old read register */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+            OldRead = IOReadB(VGA_GC_DATA);
+
+            /* Loop and read from all four color bitplanes */
+            for (CurrentPlane = 0; CurrentPlane < 4; CurrentPlane++)
+            {
+                BYTE MemoryValue;
+
+                IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+                IOWriteB(VGA_GC_DATA, CurrentPlane);
+
+                EmulatorReadMemory(&EmulatorContext, Address, &MemoryValue, sizeof(BYTE));
+                Value |= ((MemoryValue >> Shift) & 1) << CurrentPlane;
+            }
+
+            /* Restore old read register */
+            IOWriteB(VGA_GC_INDEX, VGA_GC_READ_MAP_SEL_REG);
+            IOWriteB(VGA_GC_DATA, OldRead);
+            return Value;
+        }
+
+        case 0x11: /* 640x480 2c */
+        {
+            BYTE MemoryValue;
+            BYTE ValueShift = 7 - X % 8;
```

Comment:
```suggestion
            BYTE ValueShift = 7 - (X & 7);
```
(btw. it would be interesting to see how this performs compared to equivalent `~X & 7` .)

Reply:
A [quick check](https://godbolt.org/z/4rzqGM) in Compiler Explorer suggests that GCC at around the same version as what's in RosBE outputs the same code for both. MSVC on the other hand outputs a bit faster assembly for `~X & 7` even with optimizations enabled.